python/revng/tupletree/__init__.py

```diff
@@ -370,3 +388,374 @@ def init_check_kw_only(self, *args, **kwargs):
         original_init(self, *args, **kwargs)
 
     base_class.__init__ = init_check_kw_only  # type: ignore
+
+
+def construct_type(type_info: dict, data):
+    # if data not `str` or `dict` then its already constructed so we just return it.
+    if not isinstance(data, str) and not isinstance(data, dict):
+        return data
+
+    if type_info["ctor"] == "enum":
+        return type_info["possibleValues"](data)
+    elif type_info["ctor"] == "native":
+        return type_info["type"](data)
+    if type_info["ctor"] == "parse":
+        if type_info["type"].__name__ in ("TypeDefinition", "Type"):
+            return type_info["type"].from_dict(**data)
+        return type_info["type"](data)
+    elif type_info["ctor"] == "class":
+        return type_info["type"].from_dict(**data)
+
+
+class DiffSet:
+    def __init__(self, changes):
+        self.changes = changes
+
+    @classmethod
+    def from_dict(cls, **kwargs):
+        changes = []
+        for diff in kwargs["Changes"]:
+            changes.append(Diff(diff["Path"], diff.get("Add"), diff.get("Remove")))
+        return DiffSet(changes)
+
+    @classmethod
+    def yaml_constructor(cls, loader, node):
+        mapping = loader.construct_mapping(node, deep=True)
+        return cls.from_dict(**mapping)
+
+    @classmethod
+    def yaml_representer(cls, dumper: YamlDumper, instance) -> yaml.Node:
+        return dumper.represent_dict({"Changes": instance.changes})
+
+    def __repr__(self):
+        return repr({"Changes": self.changes})
+
+    def __str__(self):
+        return str({"Changes": self.changes})
+
+
+class Diff:
+    def __init__(self, path, add, remove):
+        self.Path = path
+        self.Add = add
+        self.Remove = remove
+
+    @classmethod
+    def yaml_representer(cls, dumper: YamlDumper, instance):
+        return dumper.represent_dict(instance._to_dict())
+
+    def _to_dict(self):
+        d = {"Path": self.Path}
+        if self.Add:
+            d["Add"] = self.Add
+        if self.Remove:
+            d["Remove"] = self.Remove
+        return d
+
+    def __repr__(self):
+        return repr(self._to_dict())
+
+    def __str__(self):
+        return str(self._to_dict())
+
+    def is_valid(self):
+        return self.Add is not None or self.Remove is not None
+
+
+YamlDumper.add_representer(Diff, Diff.yaml_representer)
+
+
+def _parse_diff(diffs: DiffSet, get_type_info: Callable) -> DiffSet:
```

Comment:
I removed the function and instead called the tupletree function directly in [`load_diff`](https://github.com/revng/revng/pull/398/commits/f9733d6ecc4abbd0f82687c8f8f97e52a395a468#diff-2939980024330b4bb7d052a4eac10c83f07d4e23fbfcd5b5be010aad6df97df5R78) since this is the only occurrence that uses the _parse_diff function